File: code_samples/section7/problems/problem7_5_2/problem7_5_2.js
/**
 * countUniquePairs
 *
 * Purpose:
 *   Counts the number of *unique value pairs* {a, b} such that:
 *       a + b === target
 *
 *   Uniqueness is by VALUE, not by index.
 *   That means duplicates in the input array do not create extra pairs
 *   if they represent the same two numbers.
 *
 * Examples:
 *   nums   = [1, 1, 2, 2, 3, 3], target = 4
 *   Valid unique pairs: {1,3}, {2,2} → result = 2
 *
 *   nums   = [1, 5, 1, 5], target = 6
 *   Valid unique pair: {1,5} → result = 1
 *
 * Parameters:
 *   nums   - Array of integers (may contain duplicates).
 *   target - Target sum.
 *
 * Returns:
 *   The number of distinct value pairs that sum to the target.
 *
 * Algorithm (single pass with sets):
 *   - Use `seen` to store values we have already processed.
 *   - Use `pairs` to store unique normalized pairs.
 *
 *   For each value x in nums:
 *     1) Compute y = target - x.
 *     2) If y is already in `seen`, then {x, y} is a valid pair.
 *     3) Normalize the pair so order does not matter:
 *          a = min(x, y), b = max(x, y)
 *     4) Encode the pair as a string "a,b" and insert into `pairs`.
 *     5) Add x to `seen`.
 *
 * Why normalization matters:
 *   - Without it, {1,3} and {3,1} would be treated as different pairs.
 *   - Enforcing a <= b guarantees one canonical representation.
 *
 * Complexity (expected / average):
 *   - Time:  O(n)
 *       Each element is processed once, with O(1) average-time Set operations.
 *       String creation adds some overhead but is acceptable for learning/demo.
 *   - Space: O(n)
 *       In the worst case, both sets may grow linearly.
 *
 * Notes / potential improvements:
 *   - Using strings ("a,b") is simple but not optimal.
 *     A more efficient alternative would be to store pairs as:
 *       - a custom object with a hash, or
 *       - a single packed integer / BigInt key.
 */
function countUniquePairs(nums, target) {
    // Set of values we have already seen while scanning left-to-right
    const seen = new Set();

    // Set of unique normalized pair keys ("a,b")
    const pairs = new Set();

    // Iterate through the array
    for (const x of nums) {
        // Complement needed to reach the target
        const y = target - x;

        // If y has already appeared, x and y form a valid pair
        if (seen.has(y)) {
            // Normalize order so (x,y) and (y,x) are treated the same
            const a = Math.min(x, y);
            const b = Math.max(x, y);

            // Encode normalized pair into a string key
            pairs.add(a + "," + b);
        }

        // Record current value as seen
        seen.add(x);
    }

    // Number of unique value pairs found
    return pairs.size;
}

// ============================
// Tests (top-level)
// ============================

/**
 * printArray
 *
 * Helper function to format arrays consistently.
 */
function printArray(arr) {
    return "[" + arr.join(", ") + "]";
}

// Test cases covering:
//  - duplicates that form multiple unique pairs
//  - standard two-sum case
//  - repeated single value case
//  - no-solution case
//  - same pair appearing multiple times
const tests = [
    { nums: [1, 1, 2, 2, 3, 3], target: 4, expected: 2, note: "{1,3} & {2,2}" },
    { nums: [2, 7, 11, 15],     target: 9, expected: 1, note: "{2,7}" },
    { nums: [3, 3, 3],          target: 6, expected: 1, note: "{3,3}" },
    { nums: [1, 2, 3, 4],       target: 10, expected: 0, note: "none" },
    { nums: [1, 5, 1, 5],       target: 6, expected: 1, note: "{1,5}" }
];

console.log("=== Test: countUniquePairs ===\n");

// Run each test case
for (const t of tests) {
    const result = countUniquePairs(t.nums, t.target);

    console.log(`Input nums = ${printArray(t.nums)}, target = ${t.target}`);
    console.log(`Result:   ${result}`);
    console.log(`Expected: ${t.expected} (${t.note})\n`);
}

File: code_samples/section7/problems/problem7_5_2/problem7_5_2.cpp
#include <iostream>
#include <unordered_set>
#include <vector>
#include <string>
using namespace std;

/**
 * countUniquePairs
 *
 * Purpose:
 *   Counts the number of *unique value pairs* {a, b} such that:
 *       a + b == target
 *
 *   Uniqueness is by VALUE, not by index.
 *
 * Example:
 *   nums   = [1, 1, 2, 2, 3, 3]
 *   target = 4
 *   Valid value pairs are {1,3} and {2,2} -> answer is 2.
 *
 * Parameters:
 *   nums   - input numbers (may contain duplicates, may be unsorted)
 *   target - the desired sum
 *
 * Returns:
 *   The number of distinct value pairs that sum to target.
 *
 * Approach:
 *   - Use `seen` to store numbers we've already processed.
 *   - For each x in nums:
 *       y = target - x
 *       If y is in seen, then (x,y) forms a pair.
 *       Normalize pair order as (a=min(x,y), b=max(x,y)) so order doesn't matter.
 *       Convert that normalized pair into a unique key string "a,b" and insert
 *       it into the `pairs` set to deduplicate.
 *
 * Why we need both sets:
 *   - `seen` prevents us from requiring two passes and ensures we only match
 *     against values that appeared earlier.
 *   - `pairs` ensures duplicates in nums do not inflate the count.
 *
 * Complexity (expected / average):
 *   - Time:  O(n) average for hash lookups/inserts, but with extra overhead:
 *       Creating strings for each found pair adds extra work (and allocations).
 *   - Space: O(n) for the sets.
 *
 * Notes / potential improvements:
 *   - Using string keys is simple and readable, but it is not the most efficient
 *     approach (string creation + hashing).
 *   - A more efficient alternative is to store pairs as a struct with a custom
 *     hash (or pack into 64 bits) in an unordered_set.
 */
int countUniquePairs(const vector<int> &nums, int target) {
    // Stores values we've already seen while scanning left-to-right
    unordered_set<int> seen;

    // Stores unique normalized pair identifiers ("a,b") so duplicates are ignored
    unordered_set<string> pairs;

    // Process each number in the input
    for (int x : nums) {
        // Complement needed to reach the target
        int y = target - x;

        // If we've already seen y, then x and y can form a valid pair
        if (seen.count(y)) {
            // Normalize order so (x,y) and (y,x) are treated the same
            int a = min(x, y);
            int b = max(x, y);

            // Encode the normalized pair into a string key.
            // Example: a=1, b=5 -> "1,5"
            //
            // Because we normalize (a<=b), we avoid double-counting.
            string key = to_string(a) + "," + to_string(b);

            // Insert into set; duplicates are automatically ignored
            pairs.insert(key);
        }

        // Record x as seen for future elements
        seen.insert(x);
    }

    // Number of unique normalized pairs found
    return (int)pairs.size();
}

/**
 * printVec
 *
 * Helper function to print a vector<int> in [a, b, c] format.
 */
static void printVec(const vector<int>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); i++) {
        cout << v[i];
        if (i + 1 < v.size()) cout << ", ";
    }
    cout << "]";
}

/**
 * main
 *
 * Simple console-based test harness for countUniquePairs().
 * Demonstrates:
 *   - duplicates that create only a few unique value pairs
 *   - standard two-sum case
 *   - repeated single value case (3+3)
 *   - no-solution case
 *   - duplicates arranged differently (1,5 repeated)
 */
int main() {
    cout << "=== Test: countUniquePairs ===\n\n";

    /**
     * TestCase
     *
     * Bundles one test input with its expected output and a short description
     * of what unique pair(s) should exist.
     */
    struct TestCase {
        vector<int> nums;
        int target;
        int expected;
        string description;
    };

    // Collection of test cases
    vector<TestCase> tests = {
        { {1,1,2,2,3,3}, 4, 2, "{1,3} and {2,2}" },
        { {2,7,11,15},   9, 1, "{2,7}" },
        { {3,3,3},       6, 1, "{3,3}" },
        { {1,2,3,4},    10, 0, "no pairs" },
        { {1,5,1,5},     6, 1, "{1,5}" },
    };

    // Run each test case
    for (const auto& t : tests) {
        cout << "Input: nums = ";
        printVec(t.nums);
        cout << ", target = " << t.target << "\n";

        // Execute the function under test
        int result = countUniquePairs(t.nums, t.target);

        // Print results
        cout << "Result:   " << result << "\n";
        cout << "Expected: " << t.expected
             << " (" << t.description << ")\n\n";
    }

    return 0;
}
